### PR TITLE
feat: [#495] Centralize Error Messages in Framework (PR#2) 

### PR DESCRIPTION
**Pull request by kkumar-gcc**
## 📑 Description

Closes https://github.com/goravel/goravel/issues/495

## Release Notes

- **New Features**
	- Enhanced error handling across various components, utilizing a standardized error package for improved clarity and maintainability.
  
- **Bug Fixes**
	- Improved robustness in cache and cryptographic functionalities by ensuring necessary components are validated before use.

- **Documentation**
	- Updated comments for clarity and consistency in the cache memory management methods.

- **Refactor**
	- Simplified error reporting mechanisms in database and cache management, promoting a more dynamic and structured approach.

## ✅ Checks

- [ ] Added test cases for my code

### CODE REVIEW

**Comment by kkumar-gcc** [comment]
@coderabbitai review

File: auth/service_provider.go [hunk 1]
```diff
@@ -19,8 +20,21 @@
...
+		}
+
+		ormFacade := app.MakeOrm()
+		if ormFacade == nil {
+			return nil, errors.ErrOrmFacadeNotSet.SetModule(errors.ModuleAuth)
+		}
+
 		return NewAuth(config.GetString("auth.defaults.guard"),
-			app.MakeCache(), config, parameters["ctx"].(http.Context), app.MakeOrm()), nil
+			cacheFacade, config, parameters["ctx"].(http.Context), ormFacade), nil
```
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply auth/service_provider.go:37]: cc @kkumar-gcc

File: cache/application.go [hunk 2]
```diff
@@ -9,7 +9,6 @@ import (
 type Application struct {
 	cache.Driver
 	config config.Config
-	driver Driver
```
Comment (hwbrzzl) [inline cache/application.go:12]: We can remove the Driver interface, then change `DriverImpl` to `Driver`, change here to `driver * Driver`, and keep `instance, err := app.driver.New(name)` in line 38.

File: crypt/aes.go [hunk 3]
```diff
@@ -5,12 +5,12 @@ import (
 	"crypto/cipher"
 	"crypto/rand"
 	"encoding/base64"
-	"errors"
-	"fmt"
+
```
Comment (hwbrzzl) [inline crypt/aes.go:8]:
```suggestion
```

File: database/gorm/logger.go [hunk 4]
```diff
@@ -12,6 +11,8 @@ import (
 	"time"
 
 	"gorm.io/gorm/logger"
+
+	"github.com/goravel/framework/errors"
```
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply database/gorm/logger.go:15]: cc @kkumar-gcc

File: database/gorm/query_test.go [hunk 5]
```diff
@@ -1848,7 +1848,7 @@ func (s *QueryTestSuite) TestFirstOrFail() {
 				name: "fail",
 				setup: func() {
 					var user User
-					s.Equal(orm.ErrRecordNotFound, query.Query().Where("name", "first_or_fail_user").FirstOrFail(&user))
+					s.Equal(errors.ErrOrmRecordNotFound, query.Query().Where("name", "first_or_fail_user").FirstOrFail(&user))
```
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply database/gorm/query_test.go:1851]: cc @kkumar-gcc

File: errors/list.go
Comment (hwbrzzl) [inline errors/list.go:1]: Sort them.

File: errors/modules.go
Comment (hwbrzzl) [inline errors/modules.go:1]: Sort them.

File: foundation/application_test.go [hunk 6]
```diff
@@ -282,6 +282,15 @@ func (s *ApplicationTestSuite) TestMakeLang() {
 }
 
 func (s *ApplicationTestSuite) TestMakeLog() {
+	mockConfig := &mocksconfig.Config{}
```
Comment (hwbrzzl) [inline foundation/application_test.go:285]:
```suggestion
	mockConfig := mocksconfig.NewConfig()
```

Use the New* method, we don't need to call `AssertExpectations` again.

File: grpc/application.go [hunk 7]
```diff
@@ -33,20 +33,20 @@ func (app *Application) Server() *grpc.Server {
...
 	host := app.config.GetString(fmt.Sprintf("grpc.clients.%s.host", name))
 	if host == "" {
-		return nil, errors.New("client host can't be empty")
+		return nil, errors.ErrGrpcEmptyClientHost
 	}
 	if !strings.Contains(host, ":") {
 		port := app.config.GetString(fmt.Sprintf("grpc.clients.%s.port", name))
 		if port == "" {
-			return nil, errors.New("client port can't be empty")
+			return nil, errors.ErrGrpcEmptyClientPost
```
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply grpc/application.go:41]: cc @kkumar-gcc

File: schedule/service_provider.go [hunk 8]
```diff
@@ -12,7 +13,26 @@
 func (receiver *ServiceProvider) Register(app foundation.Application) {
 	app.Singleton(Binding, func(app foundation.Application) (any, error) {
 		config := app.MakeConfig()
-		return NewApplication(app.MakeArtisan(), app.MakeCache(), app.MakeLog(), config.GetBool("app.debug")), nil
+		if config == nil {
+			return nil, errors.ErrScheduleFacadeNotSet.SetModule(errors.ModuleSchedule)
+		}
```
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply schedule/service_provider.go:18]: cc @kkumar-gcc

**Review by hwbrzzl — COMMENTED** [review]
Awesome!

File: database/gorm/logger.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply database/gorm/logger.go:15]: it's actually using `Is` method from new errors module.

**Review by hwbrzzl — APPROVED** [review]

**Comment by hwbrzzl** [comment]
Looks good to me, please  check all comments from AI, resolve them if no changes are needed. And when the PR is ready to be reviewed, please change its tatus to open from draft.

**Review by hwbrzzl — APPROVED** [review]
Great 👍

**Review by devhaozi — APPROVED** [review]
LGTM

**Pull request merged**